[Flaky test] TestCluster_Run/Local_Node_matches_ExternalIPPool_nodeSelectors

**Describe the bug**

I have seen test `TestCluster_Run/Local_Node_matches_ExternalIPPool_nodeSelectors` fail in the Github CI, when run on Windows, for which the VMs are significantly slower. This is one of the unit test for the code in charge of assigning Nodes responsible for each "external" IP (e.g., Egress IP). Looking at the test, it seems that races are definitely possible: the test considers that an IP is done processing if the `ExternalIPPool` CR is returned by the lister.

https://github.com/antrea-io/antrea/blob/bd82ef668efc43c733566edefbf83ec0461e61ef/pkg/agent/memberlist/cluster_test.go#L155-L158

The only reason why the test doesn't fail more often IMO is because we use `wait.Poll` instead of `wait.PollImmediate`, with a pretty long initial sleep of 100ms.

```
2022-04-18T19:33:08.0366871Z I0418 19:33:05.397118    6972 cluster.go:321] "Starting" controllerName="MemberListCluster"
2022-04-18T19:33:08.0368219Z I0418 19:33:05.447112    6972 shared_informer.go:240] Waiting for caches to sync for MemberListCluster
2022-04-18T19:33:08.0370182Z I0418 19:33:05.447112    6972 shared_informer.go:247] Caches are synced for MemberListCluster 
2022-04-18T19:33:08.0384826Z E0418 19:33:05.447112    6972 cluster.go:451] "Processing Node event, not found" err="node \"localNodeName\" not found" eventType=0
2022-04-18T19:33:08.0388781Z I0418 19:33:05.610961    6972 cluster.go:345] "Shutting down" controllerName="MemberListCluster"
2022-04-18T19:33:08.0442702Z I0418 19:33:05.786120    6972 cluster.go:321] "Starting" controllerName="MemberListCluster"
2022-04-18T19:33:08.0445058Z I0418 19:33:05.786120    6972 shared_informer.go:240] Waiting for caches to sync for MemberListCluster
2022-04-18T19:33:08.0446136Z I0418 19:33:05.786120    6972 shared_informer.go:247] Caches are synced for MemberListCluster 
2022-04-18T19:33:08.0447279Z E0418 19:33:05.786120    6972 cluster.go:451] "Processing Node event, not found" err="node \"localNodeName\" not found" eventType=0
2022-04-18T19:33:08.0448242Z I0418 19:33:05.897999    6972 cluster.go:345] "Shutting down" controllerName="MemberListCluster"
2022-04-18T19:33:08.0449887Z --- FAIL: TestCluster_Run (0.64s)
2022-04-18T19:33:08.0452753Z     --- FAIL: TestCluster_Run/Local_Node_matches_ExternalIPPool_nodeSelectors (0.36s)
2022-04-18T19:33:08.0493261Z         cluster_test.go:168: 
2022-04-18T19:33:08.0494849Z             	Error Trace:	cluster_test.go:168
2022-04-18T19:33:08.0617444Z             	Error:      	Not equal: 
2022-04-18T19:33:08.0618944Z             	            	expected: true
2022-04-18T19:33:08.0619600Z             	            	actual  : false
2022-04-18T19:33:08.0623227Z             	Test:       	TestCluster_Run/Local_Node_matches_ExternalIPPool_nodeSelectors
2022-04-18T19:33:08.0624321Z             	Messages:   	select Node for Egress result not match
2022-04-18T19:33:08.0628451Z I0418 19:33:06.172631    6972 cluster.go:321] "Starting" controllerName="MemberListCluster"
2022-04-18T19:33:08.0689522Z I0418 19:33:06.172631    6972 shared_informer.go:240] Waiting for caches to sync for MemberListCluster
2022-04-18T19:33:08.0690780Z I0418 19:33:06.172631    6972 shared_informer.go:247] Caches are synced for MemberListCluster 
2022-04-18T19:33:08.0820223Z I0418 19:33:06.182724    6972 cluster.go:459] "Processed Node event" eventType=Join nodeName="localNodeName" affectedExternalIPPoolNum=0
2022-04-18T19:33:08.0824536Z I0418 19:33:07.790471    6972 cluster.go:461] "Processed Node event" eventType=Update nodeName="fakeNodeNameUpdate"
2022-04-18T19:33:08.0827541Z I0418 19:33:07.790993    6972 cluster.go:459] "Processed Node event" eventType=Leave nodeName="localNodeName" affectedExternalIPPoolNum=0
2022-04-18T19:33:08.0829652Z E0418 19:33:07.790993    6972 cluster.go:451] "Processing Node event, not found" err="node \"fakeNodeNameLeave\" not found" eventType=1
2022-04-18T19:33:08.0895197Z I0418 19:33:07.790993    6972 cluster.go:459] "Processed Node event" eventType=Join nodeName="localNodeName" affectedExternalIPPoolNum=0
2022-04-18T19:33:08.1053921Z E0418 19:33:07.790993    6972 cluster.go:197] "Processing Node CREATE event error" err="obtain IP addresses from K8s Node failed: 'x' is not a valid IP address" nodeName="fakeNode0"
2022-04-18T19:33:08.1055997Z I0418 19:33:07.890875    6972 cluster.go:345] "Shutting down" controllerName="MemberListCluster"
2022-04-18T19:33:08.1056443Z FAIL
2022-04-18T19:33:08.1056998Z FAIL	antrea.io/antrea/pkg/agent/memberlist	2.895s
```

--- a/pkg/agent/memberlist/cluster_test.go
+++ b/pkg/agent/memberlist/cluster_test.go
@@ -18,7 +18,6 @@ import (
 	"context"
 	"fmt"
 	"net"
-	"reflect"
 	"testing"
 	"time"
 
@@ -152,20 +151,15 @@ func TestCluster_Run(t *testing.T) {
 
 			go fakeCluster.cluster.Run(stopCh)
 
-			assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
-				newEIP, _ := fakeCluster.cluster.externalIPPoolLister.Get(eip.Name)
-				return reflect.DeepEqual(newEIP, eip), nil
-			}))
-
 			tCase.egress.Spec.ExternalIPPool = eip.Name
-			res, err := fakeCluster.cluster.ShouldSelectIP(tCase.egress.Spec.EgressIP, eip.Name)
-			// Cluster should hold the same consistent hash ring for each ExternalIPPool.
-			assert.NoError(t, err)
+			assert.NoError(t, wait.Poll(100*time.Millisecond, time.Second, func() (done bool, err error) {
+				res, err := fakeCluster.cluster.ShouldSelectIP(tCase.egress.Spec.EgressIP, eip.Name)
+				return err == nil && res == tCase.expectEgressSelectResult, nil
+			}), "select Node result for Egress does not match")
 			allMembers, err := fakeCluster.cluster.allClusterMembers()
 			assert.NoError(t, err)
-			assert.Equal(t, 1, len(allMembers), "expected Node member num is 1")
+			assert.Len(t, allMembers, 1, "expected Node member num is 1")
 			assert.Equal(t, 1, fakeCluster.cluster.mList.NumMembers(), "expected alive Node num is 1")
-			assert.Equal(t, tCase.expectEgressSelectResult, res, "select Node for Egress result not match")
 		})
 	}
 }
@@ -207,11 +201,6 @@ func TestCluster_RunClusterEvents(t *testing.T) {
 
 	go fakeCluster.cluster.Run(stopCh)
 
-	assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
-		newEIP, _ := fakeCluster.cluster.externalIPPoolLister.Get(fakeEIP1.Name)
-		return reflect.DeepEqual(newEIP, fakeEIP1), nil
-	}))
-
 	// Test updating Node labels.
 	testCasesUpdateNode := []struct {
 		name                     string
@@ -249,18 +238,15 @@ func TestCluster_RunClusterEvents(t *testing.T) {
 		if err != nil {
 			t.Fatalf("Update Node error: %v", err)
 		}
-		assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
-			newNode, _ := fakeCluster.cluster.nodeLister.Get(node.Name)
-			return reflect.DeepEqual(node, newNode), nil
-		}))
 	}
 	for _, tCase := range testCasesUpdateNode {
 		t.Run(tCase.name, func(t *testing.T) {
 			localNode.Labels = tCase.newNodeLabels
 			updateNode(localNode)
-			res, err := fakeCluster.cluster.ShouldSelectIP(tCase.egress.Spec.EgressIP, tCase.egress.Spec.ExternalIPPool)
-			assert.NoError(t, err)
-			assert.Equal(t, tCase.expectEgressSelectResult, res, "select Node for Egress result not match")
+			assert.NoError(t, wait.Poll(100*time.Millisecond, time.Second, func() (done bool, err error) {
+				res, err := fakeCluster.cluster.ShouldSelectIP(tCase.egress.Spec.EgressIP, tCase.egress.Spec.ExternalIPPool)
+				return err == nil && res == tCase.expectEgressSelectResult, nil
+			}), "select Node result for Egress does not match")
 		})
 	}
 
@@ -300,13 +286,10 @@ func TestCluster_RunClusterEvents(t *testing.T) {
 			if err != nil {
 				t.Fatalf("Update ExternalIPPool error: %v", err)
 			}
-			assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
-				newEIP, _ := fakeCluster.cluster.externalIPPoolLister.Get(fakeEIP1.Name)
-				return reflect.DeepEqual(fakeEIP1, newEIP), nil
-			}))
-			res, err := fakeCluster.cluster.ShouldSelectIP(fakeEgress1.Spec.EgressIP, fakeEgress1.Spec.ExternalIPPool)
-			assert.NoError(t, err)
-			assert.Equal(t, tCase.expectEgressSelectResult, res, "select Node for Egress result not match")
+			assert.NoError(t, wait.Poll(100*time.Millisecond, time.Second, func() (done bool, err error) {
+				res, err := fakeCluster.cluster.ShouldSelectIP(fakeEgress1.Spec.EgressIP, fakeEgress1.Spec.ExternalIPPool)
+				return err == nil && res == tCase.expectEgressSelectResult, nil
+			}), "select Node result for Egress does not match")
 		})
 	}
 
@@ -326,16 +309,15 @@ func TestCluster_RunClusterEvents(t *testing.T) {
 		Spec:       crdv1a2.EgressSpec{ExternalIPPool: fakeEIP2.Name, EgressIP: fakeEgressIP2},
 	}
 	assert.NoError(t, createExternalIPPool(fakeCluster.crdClient, fakeEIP2))
-	assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
-		newEIP, _ := fakeCluster.cluster.externalIPPoolLister.Get(fakeEIP2.Name)
-		return reflect.DeepEqual(newEIP, fakeEIP2), nil
-	}))
+
 	assertEgressSelectResult := func(egress *crdv1a2.Egress, expectedRes bool, hasSyncedErr bool) {
-		res, err := fakeCluster.cluster.ShouldSelectIP(egress.Spec.EgressIP, egress.Spec.ExternalIPPool)
-		if !hasSyncedErr {
-			assert.NoError(t, err)
-		}
-		assert.Equal(t, expectedRes, res, "select Node for Egress result not match")
+		assert.NoErrorf(t, wait.Poll(100*time.Millisecond, time.Second, func() (done bool, err error) {
+			res, err := fakeCluster.cluster.ShouldSelectIP(egress.Spec.EgressIP, egress.Spec.ExternalIPPool)
+			if hasSyncedErr {
+				return err != nil, nil
+			}
+			return err == nil && res == expectedRes, nil
+		}), "select Node result for Egress '%s' does not match", egress.Name)
 	}
 	assertEgressSelectResult(fakeEgress2, true, false)
 	assertEgressSelectResult(fakeEgress1, false, false)
@@ -346,10 +328,6 @@ func TestCluster_RunClusterEvents(t *testing.T) {
 		if err != nil {
 			t.Fatalf("Delete ExternalIPPool error: %v", err)
 		}
-		assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
-			newEIP, _ := fakeCluster.cluster.externalIPPoolLister.Get(eipName)
-			return nil == newEIP, nil
-		}))
 	}
 	deleteExternalIPPool(fakeEIP2.Name)
 	assertEgressSelectResult(fakeEgress2, false, true)
@@ -390,10 +368,6 @@ func TestCluster_RunClusterEvents(t *testing.T) {
 		if err != nil {
 			t.Fatalf("Delete Node error: %v", err)
 		}
-		assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
-			newNode, _ := fakeCluster.cluster.nodeLister.Get(node.Name)
-			return nil == newNode, nil
-		}))
 	}
 	deleteNode(localNode)
 	assertEgressSelectResult(fakeEgress2, false, true)

